packages/firestore/src/model/mutation.ts

```diff
@@ -515,35 +600,57 @@ function applyPatchMutationToRemoteDocument(
   maybeDoc: MaybeDocument | null,
   mutationResult: MutationResult
 ): MaybeDocument {
-  debugAssert(
-    mutationResult.transformResults == null,
-    'Transform results received by PatchMutation.'
-  );
-
   if (!preconditionIsValidForDocument(mutation.precondition, maybeDoc)) {
-    // Since the mutation was not rejected, we know that the  precondition
+    // Since the mutation was not rejected, we know that the precondition
     // matched on the backend. We therefore must not have the expected version
     // of the document in our cache and return an UnknownDocument with the
     // known updateTime.
     return new UnknownDocument(mutation.key, mutationResult.version);
   }
 
-  const newData = patchDocument(mutation, maybeDoc);
+  let newData = patchDocument(mutation, maybeDoc);
+  if (mutation.fieldTransforms && mutationResult.transformResults) {
+    const transformResults = serverTransformResults(
+      mutation.fieldTransforms,
+      maybeDoc,
+      mutationResult.transformResults
+    );
+    newData = transformObject(
+      mutation.fieldTransforms,
+      newData,
+      transformResults
+    );
```

Comment:
We might be able to reduce code duplication if you pass `transformResults` to `patchDocument ` and let it before the duties of `transformObject` as well.